marimo/_server/api/endpoints/ws.py

```diff
@@ -95,6 +106,118 @@ async def websocket_endpoint(
     ).start()
 
 
+@dataclass
+class YCell:
+    ydoc: Doc
+    clients: int
+    cleaner: asyncio.Task[None] | None = None
+
+
+ycells: dict[str, YCell] = {}
+ycell_lock = asyncio.Lock()
+
+
+def put_updates(
+    update_queue: asyncio.Queue[bytes], event: TransactionEvent
+) -> None:
+    update = event.update
+    update_queue.put_nowait(update)
+
+
+async def send_updates(
+    websocket: WebSocket,
+    update_queue: asyncio.Queue[bytes],
+    cell_id: str,
+) -> None:
+    try:
+        while True:
+            update = await update_queue.get()
+            message = create_update_message(update)
+            await websocket.send_bytes(message)
+    except Exception:
+        LOGGER.debug(
+            f"Could not send Y update to client for cell with ID {cell_id}",
+        )
+
+
+async def clean_cell(cell_id: str) -> None:
+    # Clients disconnect/reconnect to keep the WebSocket connection alive,
+    # or because of network issues.
+    # Keep the ycell in memory for one minute after client disconnects,
+    # so that it can be synchronized and to prevent content duplication.
+    # This task will be cancelled when any client reconnets.
+    await asyncio.sleep(60)
+    async with ycell_lock:
+        ycell = ycells[cell_id]
+        if not ycell.clients:
+            del ycells[cell_id]
+
+
+@router.websocket("/ws/{cell_id}")
```

Comment:
Do you expect there to be scalability concerns if there are potentially 40+ cells/WebSocket connections?

Reply:
Hard to say, if that happens the next step would be to have this top-level shared document for the whole notebook.